rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,264 @@
+ ```
+ dst_f32 = conv(src_f32, wei_f32)
+
+ src_f32 = src_s8 * src_scale_f32
+ wei_f32 = wei_s8 * wei_scale_f32
+ dst_f32 = dst_s8 * dst_scale_f32
+
+ output_scale_f32 = src_scale_f32 * wei_scale_f32 / dst_scale_f32
+ acc_s32 = conv(src_s8, wei_s8)
+ dst_s8 = acc_s32 * output_scale_f32
+ ```
+ - Argument scaling. Operation's inputs have different scales and they should
+ be taken into account while computing the operation. For example:
+ ```
+ dst_f32 = src0_f32 + src1_f32
+
+ src0_f32 = src0_s8 * src0_scale_f32
+ src1_f32 = src1_s8 * src1_scale_f32
+ dst_f32 = dst_s8 * dst_scale_f32
+
+ # scales src0_scale_f32 and src1_scale_f32 are required to compute operation
+ correctly.
+ acc_s32 = src0_s8 + src1_s8 # doesn't work if scales are different
+ dst_s8 = acc_s32 * scale_f32
+ ```
+
+The following features make scaling more flexible:
+- runtime support. Allows model to set scale at runtime;
+- broadcasting/mask. Allows model to use multiple scales.
+
+
+### 1.3. Quantization scaling in oneDNN
+
+oneDNN has multiple mechanisms to support output and argument scaling:
+
+| mechanism                      | argument  | compile-time | run-time | mask |
+|--------------------------------|-----------|--------------|----------|------|
+| attr::output_scale             | dst       | yes          | yes      | yes  |
+| attr::scales                   | all       | yes          | yes      | yes  |
+| attr::p_op::binary_mul         | dst       | no           | yes      | yes  |
+| attr::rnn_data_qparams         | src & dst | yes          | no       | no   |
+| attr::rnn_weights_qparams      | wei       | yes          | no       | no   |
+| attr::p_op::built-in parameter | dst       | yes          | no       | no   |
+
+
+The following table shows scaling mechanisms support by primitives API:
+
+| primitive                 | mechanism                   |
+|---------------------------|-----------------------------|
+| Conv/IP/Matmul            | attr::output_scale          |
+|                           | attr::p_op::binary_mul      |
+| RNN                       | attr::rnn_data_qparams      |
+|                           | attr::rnn_weights_qparams   |
+| Batch/Layer normalization |                             |
+| Binary                    | attr::scales                |
+| attr::p_op::binary        |                             |
+| Concat                    |                             |
+| Eltwise                   |                             |
+| attr::p_op::eltwise       | built-in parameter          |
+| LRN                       |                             |
+| (Log)Softmax              | attr::output_scale[[#2]][2] |
+| Pooling                   | attr::p_op::binary_mul      |
+| PRelu                     |                             |
+| Resampling                | attr::p_op::binary_mul      |
+| Shuffle                   |                             |
+| Sum                       |                             |
```

Comment:
There seem to be one more kind of scales - scales in primitive descriptor (sum): https://github.com/oneapi-src/oneDNN/blob/c0c5582d671bf573a576f4e63b2edeb02fbd0f8a/include/oneapi/dnnl/dnnl.hpp#L4427-L4430

I think removing this would be the right thing to do as well. What do you think?

Reply:
Indeed there is. As far as I remember these scales are not quantization scales but sum scales that are used for all data types, so I would keep them out of scope of this RFC.

Reply:
Added to the RFC.